lib/services/local/plugins.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package local
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+const (
+	pluginsPrefix = "plugins"
+)
+
+// PluginsService manages plugin instances in the backend.
+type PluginsService struct {
+	backend backend.Backend
+}
+
+func NewPluginsService(backend backend.Backend) *PluginsService {
+	return &PluginsService{backend: backend}
+}
+
+// CreatePlugin implements services.Plugins
+func (s *PluginsService) CreatePlugin(ctx context.Context, req *proto.CreatePluginRequest) error {
+	if req == nil {
+		return trace.BadParameter("req is nil")
+	}
+
+	plugin := req.Plugin
+	if plugin == nil {
+		return trace.BadParameter("req.Plugin is nil")
+	}
+
+	// TODO: exchange initial credentials for live credentials
```

Comment:
```suggestion
	// TODO(justinas): exchange initial credentials for live credentials
```

What's the plan here? Any sort of active interaction with a network service (like trading some initial credential for a renewable one) should not be done here but rather in `auth.Server`, and here we should only really deal with data (perhaps by taking a `types.Plugin` rather than a `CreatePluginRequest`)

Reply:
I agree that auth being responsible for the exchange makes more sense.

I initially extended `proto.CreatePluginRequest` to the service to keep the signature the same as auth, so we can add `services.Plugins` to [ClientI](https://github.com/gravitational/teleport/blob/06bf633ab5cbeca37690b1599fa028f36c474173/lib/auth/clt.go#L1574) and the auth server implements those methods (since signatures match).

We could either:
* keep `CreatePluginRequest` in the service, but assert that the credentials have been exchanged by the caller (this keeps the methods identical and allows interface composition)
* make the service have a `CreatePlugin(types.Plugin)` (which looks cleaner), and manually define a `CreatePlugin(*proto.CreatePluginRequest)` in `ClientI`.

Reply:
I'd prefer option 2 here.

I've so far found that our attempts to keep signatures the same across layers tends to lead to more pain (and only slightly reducing duplication/boilerplate), and I think that this accepting `types.Plugin` avoids us coupling this method to much to the transport layer. This'll make life easier if we introduce a V2 format of the message in future.

This also reduces the likelihood of a refactoring snafu leading to this implementation becoming exposed without sufficient authentication of initial credentials.